provers/groth16/src/qap.rs

```diff
@@ -125,3 +65,100 @@ impl QuadraticArithmeticProgram {
         })
     }
 }
+
+impl QuadraticArithmeticProgram {
+    pub fn from_r1cs(r1cs: R1CS) -> QuadraticArithmeticProgram {
+        let num_gates = r1cs.number_of_constraints();
+        let next_power_of_two = num_gates.next_power_of_two();
+        let pad_zeroes = next_power_of_two - num_gates;
+
+        let mut l: Vec<Polynomial<FrElement>> = vec![];
+        let mut r: Vec<Polynomial<FrElement>> = vec![];
+        let mut o: Vec<Polynomial<FrElement>> = vec![];
+        for i in 0..r1cs.witness_size() {
+            let [l_poly, r_poly, o_poly] =
+                get_variable_lro_polynomials_from_r1cs(&r1cs, i, pad_zeroes);
+            l.push(l_poly);
+            r.push(r_poly);
+            o.push(o_poly);
+        }
+
+        QuadraticArithmeticProgram {
+            l,
+            r,
+            o,
+            num_of_gates: next_power_of_two,
+            num_of_public_inputs: r1cs.number_of_inputs + 1, // Compansate for "1"
```

Comment:
Typo, compansate. Also, it doesn't say why so it's not quite useful

Reply:
Unnecessary 1. My mistake. Removed.